### ISSUE TITLE
IndexError: String index out of range

### ISSUE THREAD

**Issue by sirgeorgesawcon** [Qiskit-Extensions/circuit-knitting-toolbox#487]
I am following the legacy [CutQC first tutorial](https://qiskit-extensions.github.io/circuit-knitting-toolbox/circuit_cutting/cutqc/tutorials/tutorial_1_automatic_cut_finding.html) , and I am encountering this issue if I use a different circuit from the one mentioned in the tutorial.

## How to recreate the error?
I'm using the following circuit

```python
from qiskit.circuit.library import EfficientSU2

qc = EfficientSU2(4, entanglement="linear", reps=2).decompose()
qc.assign_parameters([0.4] * len(qc.parameters), inplace=True)

qc.draw("mpl",style='iqx', scale=0.8)
```

![output](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/0d1450c9-6046-424e-a099-dcfaabadbb52)

Then I find the cut automatically, using these set of parameters:

```python
%%capture

from circuit_knitting.cutting.cutqc import cut_circuit_wires

cuts = cut_circuit_wires(
    circuit=qc,
    method="automatic",
    max_subcircuit_width=3,
    max_cuts=5,
    num_subcircuits=[2],
)
```

It gives me two subcircuits:

![2](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/96f4da2c-78a1-4319-96f7-1bb32c6b28a8)
![1](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/4e91556b-db14-44d2-8696-4e91faec2492)

Then When I try to run it using:

```python
from qiskit_ibm_runtime import QiskitRuntimeService  # noqa: F401

# Use local versions of the primitives by default.
service = None

# Uncomment the following line to instead use Qiskit Runtime Service.
# service = QiskitRuntimeService()

from qiskit_ibm_runtime import Options

# Set the Sampler and runtime options
options = Options(execution={"shots": 4000})

# Run 2 parallel qasm simulator threads
backend_names = ["ibmq_qasm_simulator"] * 2
```

and use the `evaluate_subcircuits` function

```python
from circuit_knitting.cutting.cutqc import evaluate_subcircuits

subcircuit_instance_probabilities = evaluate_subcircuits(cuts)

# Uncomment the following lines to instead use Qiskit Runtime Service as configured above.
# subcircuit_instance_probabilities = evaluate_subcircuits(cuts,
#                                                          service_args=service.active_account(),
#                                                          backend_names=backend_names,
#                                                          options=options,
#                                                         )
```

It gives me the error:

```IndexError: string index out of range```

If I use the circuit in tutorial, it works out smoothly. 

## System Info

Python : 3.10.11
Qiskit : 0.46.0
qiskit_ibm_runtime : 0.19.1
qiskit_aer :  0.13.3

**Comment by caleb-johnson** [comment]
> I am following the legacy [CutQC first tutorial](https://qiskit-extensions.github.io/circuit-knitting-toolbox/circuit_cutting/cutqc/tutorials/tutorial_1_automatic_cut_finding.html) , and I am encountering this issue if I use a different circuit from the one mentioned in the tutorial.
> 
> ## How to recreate the error?
> I'm using the following circuit
> 
> ```python
> from qiskit.circuit.library import EfficientSU2
> 
> qc = EfficientSU2(4, entanglement="linear", reps=2).decompose()
> qc.assign_parameters([0.4] * len(qc.parameters), inplace=True)
> 
> qc.draw("mpl",style='iqx', scale=0.8)
> ```
> 
> ![output](https://private-user-images.githubusercontent.com/63145298/304328019-0d1450c9-6046-424e-a099-dcfaabadbb52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzAwMzEsIm5iZiI6MTcwODUyOTczMSwicGF0aCI6Ii82MzE0NTI5OC8zMDQzMjgwMTktMGQxNDUwYzktNjA0Ni00MjRlLWEwOTktZGNmYWFiYWRiYjUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE1MzUzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxZTIyZWZkZTg0YTkxM2NjMWM5ZWI2Nzk0MzFhN2VmMDRhYzNmNzQ5YTkzNjQwYTViZWZiOWQwNzI4MjIzY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UwOYlRdaLS5cBFXluHFUAtFbJLxhCYtlm4nj99wIF3c)
> 
> Then I find the cut automatically, using these set of parameters:
> 
> ```python
> %%capture
> 
> from circuit_knitting.cutting.cutqc import cut_circuit_wires
> 
> cuts = cut_circuit_wires(
>     circuit=qc,
>     method="automatic",
>     max_subcircuit_width=3,
>     max_cuts=5,
>     num_subcircuits=[2],
> )
> ```
> 
> It gives me two subcircuits:
> 
> ![2](https://private-user-images.githubusercontent.com/63145298/304328248-96f4da2c-78a1-4319-96f7-1bb32c6b28a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzAwMzEsIm5iZiI6MTcwODUyOTczMSwicGF0aCI6Ii82MzE0NTI5OC8zMDQzMjgyNDgtOTZmNGRhMmMtNzhhMS00MzE5LTk2ZjctMWJiMzJjNmIyOGE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE1MzUzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNWExNGZlMTYzY2YyMjliZWQwMDU0NDVmNjEyMTE0ZDZlZDdkMGQyZDc4MjZiNDc0ZTdjN2JhZDQ1NjgzMjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gDf2ioYqHDrok9VpIgU8hmm7DxqsZriTQjHlgy202mk) ![1](https://private-user-images.githubusercontent.com/63145298/304328255-4e91556b-db14-44d2-8696-4e91faec2492.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzAwMzEsIm5iZiI6MTcwODUyOTczMSwicGF0aCI6Ii82MzE0NTI5OC8zMDQzMjgyNTUtNGU5MTU1NmItZGIxNC00NGQyLTg2OTYtNGU5MWZhZWMyNDkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE1MzUzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyYzJhZDBkMTllZjQ4MGJjYTRiZWZiZGExMjU3YjY0N2ZlYjFiMWUwZDlhZmVhOGRmMThmNjhhOTQxOWVlZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7k2Mb2GrNo1EJFGfPM2awLtpNaGnzGCL6W-YK0_tjFk)
> 
> Then When I try to run it using:
> 
> ```python
> from qiskit_ibm_runtime import QiskitRuntimeService  # noqa: F401
> 
> # Use local versions of the primitives by default.
> service = None
> 
> # Uncomment the following line to instead use Qiskit Runtime Service.
> # service = QiskitRuntimeService()
> 
> from qiskit_ibm_runtime import Options
> 
> # Set the Sampler and runtime options
> options = Options(execution={"shots": 4000})
> 
> # Run 2 parallel qasm simulator threads
> backend_names = ["ibmq_qasm_simulator"] * 2
> ```
> 
> and use the `evaluate_subcircuits` function
> 
> ```python
> from circuit_knitting.cutting.cutqc import evaluate_subcircuits
> 
> subcircuit_instance_probabilities = evaluate_subcircuits(cuts)
> 
> # Uncomment the following lines to instead use Qiskit Runtime Service as configured above.
> # subcircuit_instance_probabilities = evaluate_subcircuits(cuts,
> #                                                          service_args=service.active_account(),
> #                                                          backend_names=backend_names,
> #                                                          options=options,
> #                                                         )
> ```
> 
> It gives me the error:
> 
> `IndexError: string index out of range`
> 
> If I use the circuit in tutorial, it works out smoothly.
> 
> ## System Info
> Python : 3.10.11 Qiskit : 0.46.0 qiskit_ibm_runtime : 0.19.1 qiskit_aer : 0.13.3

Thanks @sirgeorgesawcon , will try to reproduce

**Comment by caleb-johnson** [comment]
@sirgeorgesawcon , we have been seeing several different problems with cutqc. It is legacy software, and we are trying to migrate users to the [newer wire cutting functionality](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/blob/main/docs/circuit_cutting/how-tos/how_to_specify_cut_wires.ipynb).

We also have an implementation of an automat cut-finder which supports both gates and wire cuts in a branch [here](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/tree/LO-circuit-cuts-optimizer), which should be available in the upcoming release.

I wonder if this new package suits your needs.

**Comment by sirgeorgesawcon** [comment]
Hi @caleb-johnson 
>> We also have an implementation of an automat cut-finder which supports both gates and wire cuts in a branch [here](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/tree/LO-circuit-cuts-optimizer), which should be available in the upcoming release.

Is this version the one that does the gate and wire cut simultaneously in a given quantum circuit? Based on [this](https://arxiv.org/abs/2308.09567) work.

**Comment by caleb-johnson** [comment]
> Is this version the one that does the gate and wire cut simultaneously in a given quantum circuit? Based on [this](https://arxiv.org/abs/2308.09567) work.

Yes, our newer `cutting` package can do both gate and wire cutting. They both use quasi-probability decompositions to decompose 2-qubit gates and wires into local operations. We drew from [the references at the bottom of this page](https://qiskit-extensions.github.io/circuit-knitting-toolbox/circuit_cutting/explanation/index.html) during implementation.

In addition to the tutorial(s) I linked above, we have an automatic cut-finder coming soon, which will help users find cut locations to most efficiently partition their circuit. An example of how it will work can be found in a branch [here](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/blob/LO-circuit-cuts-optimizer/docs/circuit_cutting/tutorials/04_automatic_cut_finding.ipynb).

Please let us know if you try out the new `cutting` package! :)

**Comment by sirgeorgesawcon** [comment]
I tried the `cutting` package, and from what I see it works with Qiskit 0.45 and less, and doesn't with Qiskit 1.0, but I saw other PR and saw that people are working on it too. Sometimes I get the error of `divide by 0` using some random circuit and some random choice of observables, but it still cuts the circuit nevertheless.

But the result from uncut circuit and the one we cut if much different, 

For instance 

```python 
import numpy as np
from qiskit.circuit.random import random_circuit
'''
This function generates a random quantum circuit, of given size and width, and deopth,
and also the max entangled operation
'''


from qiskit.quantum_info import PauliList

circuit = random_circuit(8, 5, max_operands=2, seed=1242)
observables = PauliList(["IZIIIIII", "IIIIZIII", "IIIIIIIZ"])
circuit.draw("mpl",style='iqx', scale=0.8)
```

makes this circuit 
![output](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/5a1a34c1-302a-4254-8f3c-272b6c738ff2)


and giving it these parameters to look for a cut

```python
from circuit_knitting.cutting import find_cuts

# specify the settings
optimization_settings = {"rand_seed": 42}
# specify the constraints
device_constraints = {"num_QPUs": 2, "qubits_per_QPU": 5}
# do the cut
cut_circuit = find_cuts(circuit, optimization_settings, device_constraints)
# draw the cut circuit
cut_circuit.draw("mpl",style='iqx', scale=0.8)
```

It draws the following

![output](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/b2277be1-e933-4c80-885a-d03a0fc1786e)


and after adding `ancilla`

![output1](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/assets/63145298/fe32ebda-6e09-404a-b727-b20791e39a56)



It is a nice cut, since the sampling overhead is around 58.

But, the moment I reconstruct the circuit

```python
from qiskit_aer.primitives import Sampler

# Set up a Qiskit Aer Sampler primitive for each circuit partition
samplers = {
    label: Sampler(run_options={"shots": 2**12}) for label in subexperiments.keys()
}

# Retrieve results from each partition's subexperiments
results = {
    label: sampler.run(subexperiments[label]).result()
    for label, sampler in samplers.items()
}
```

and compare it:

```python
from circuit_knitting.cutting import reconstruct_expectation_values

reconstructed_expvals = reconstruct_expectation_values(
    results,
    coefficients,
    subobservables,
)
from qiskit_aer.primitives import Estimator

estimator = Estimator(run_options={"shots": None}, approximation=True)
exact_expvals = (
    estimator.run([qc] * len(observables), list(observables)).result().values
)
print(
    f"Reconstructed expectation values: {[np.round(reconstructed_expvals[i], 8) for i in range(len(exact_expvals))]}"
)
print(
    f"Exact expectation values: {[np.round(exact_expvals[i], 8) for i in range(len(exact_expvals))]}"
)
print(
    f"Errors in estimation: {[np.round(reconstructed_expvals[i]-exact_expvals[i], 8) for i in range(len(exact_expvals))]}"
)
print(
    f"Relative errors in estimation: {[np.round((reconstructed_expvals[i]-exact_expvals[i]) / exact_expvals[i], 8) for i in range(len(exact_expvals))]}"
)
```

We get 

```
Reconstructed expectation values: [-0.56650392, 0.6264455, -0.86957383]
Exact expectation values: [0.1948505, 0.47016598, 0.66808262]
Errors in estimation: [-0.76135442, 0.15627952, -1.53765645]
Relative errors in estimation: [-3.90737728, 0.33239223, -2.30159624]
```

There's a big gap between the uncut and cut values.  It might be because I'm doing something wrong during the reconstruction part of the circuit, but I can't figure it out.



Basically I want to contribute a tutorial on solving a big MaxCut problem with QAOA, with using both gate and wire cut, and that's what I'm currently working on.  But the initial results don't match, for a random circuit.
